### PR TITLE
Adding ChiselEnum Documentation Entry

### PR DESCRIPTION
**Pull request by chrisbaldwin2**
Added documentation for the ChiselEnum type with verified examples

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- documentation

#### API Impact

No API Impact

#### Backend Code Generation Impact

No Backend Code Generation Impact

#### Desired Merge Strategy

- Squash

#### Release Notes

Added documentation for the ChiselEnum object type.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (3.2.x, 3.3.x, 3.4.0, 3.5.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by chrisbaldwin2** [comment]
Just adding some documentation for the ChiselEnum type

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:3]: Thanks for writing this!

what is a "regular Enum"? Do you mean `scala.Enum`, or something else?

Also, not sure what you mean by "allow for function names". What is a "function name"?

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:43]: shouldn't the default be one of the `ChiselEnum` types?

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:61]: What happens if the value isn't increasing? Is it a runtime error?

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:97]: the example below is a nice example of `litValue`, can you show an example of `.Type` usage?

Also, does `sel.litValue == 0.U` also work? It seems more intuitive... actually it seems way more intuitive to actually use the Enum values in such a comparison. The below seems like a pattern that should be discouraged

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:3]: By regular Enum I mean the Chisel enum. I can clarify that. 
By function name I mean like a opcode being passed into an ALU or CMP.

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:125]: Before this PR is merged let's get rid of the `I don't believe` and just make sure what is written is correct.

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:43]: It can be a default value depends on whether logic is being simplified or error checked. Something like `DontCare` or `0.U` can be used for checking that unknown selects are used. I believe that this is a matter of implementation.

Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:43]: Oh, I misread the example, the type is on the select not the data.

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:61]: An exception is thrown when Chisel emits Verilog.
> chisel3.internal.ChiselException: Exception thrown when elaborating ChiselGeneratorAnnotation

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:43]: Correct!

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:97]: .Type is used in line 100. In the testbench, the .Type function is used to pass the ChiselEnum as a parameter since we do not want the literal value!

Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:97]: `sel.litValue == 0.U` does not work since .litValue returns a BigInt and not a UInt. This block is pulled out of my testbench which tests that the ChiselEnum type is enumerating correctly and that all values match as expected.

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:109]:
```suggestion
The ChiselEnum type also has methods `.all` and `.getWidth` where `all` returns all of the enum instances and `getWidth` returns the width of the hardware type.
```

**Comment by chrisbaldwin2** [comment]
@jackkoenig do you know why this is failing to emitVerilog? I have the code building here https://scastie.scala-lang.org/4NTKq8i2Rh228cDRQzImbg but fails ci.

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:3]:
```suggestion
The ChiselEnum type can be used to reduce the chance of error when encoding mux selectors, opcodes, and functional unit operations. In contrast with`Chisel.Enum`, `ChiselEnum` can be included in `Bundle`s.
```

I still don't know what you mean by "names to be explicitly passed" or "through Bundles like IO". Even from an english language level, I don't know if "like IO" means "Bundles which are similar to IO" or "similar to the way IO can be passed through Bundles" Is the above correct?

Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:3]: Well I wanted to explicitly mention `IO` because it is the most prevalent use of `Bundle`s. When I was first trying to use `Enum` for my mux selects, I couldn't pass the `Enum` type through the `IO` of my modules. This is only possible with `ChiselEnum`s which define a `Data` type that can be passed through Bundles and IO. Since `IO` takes in a `Bundle` object, I was trying to make it clear that `ChiselEnum`s can be passed through `Bundle`s like those used in `IO` while `Enum` objects cannot

Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:3]: thanks for the explanation! What do you think of my suggested change above? (Feel free to use any/all/none of it)

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:37]: i think this is the reason your thing is failing CI, there is a bug that shows up in `mdoc` + naming plugin +` IO( new Bundle {...})`

You can try workaround this by defining your bundle type outside of the IO call:

```
class ALUMuxIO extends Bundle {
...
}
...

val io = IO(new ALUMuxIO)
```

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:26]:
```suggestion
        val selectRS1, selectPC = Value
        /**
            "selectRS1" -> 0.U,
            "selectPC"  -> 1.U
        */
```

for the purposes of a clear example, can I suggest more generic names above?

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:3]: It looks good. The only thing that I would change is that `Enum` is in the `Chisel.util` package, not in `Chisel`. I don't know if you were trying to reference the package or not.

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md]: Oh that's a pretty nasty issue. I separated the bundle class out and looks like it is passing the ci now. Thanks!

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:3]: yeah I was trying to reference the package to be explicit what we meant by "Enum". I'll fix it to say util

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:7]: can you break these up or make some of these invisible/silent since `ChiselStage` is not relevant to to "Importing the Chisel Enum module" ?

Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:7]: I'm not sure the general pattern in these docs if things that are only relevant for the examples are hidden or if the code examples are all supposed to work right out of the box

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:65]:
```suggestion
is thrown during Verilog generation.
```

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:101]: gotcha. I can't resolve this comment for some reason but you've answered my question

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:115]:
```suggestion
As of 2/26/2021, the width of the values is always inferred. To work around this, you can add an extra `Value` that forces the width that is desired. This is shown in the example below, where we add a field `ukn` to force the width to be 3 bits wide: 
```

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:7]: In referencing https://raw.githubusercontent.com/chipsalliance/chisel3/master/docs/src/explanations/naming.md they also had ChiselStage shown

Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:7]: yeah, gotcha. Then let's follow that pattern a bit more ... I'll suggest a change

Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:7]: I also believe this helps people who want to copy what is on screen and get it to run locally on their machine.

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:16]:
```suggestion

## Functionality and Examples

```scala mdoc
// Imports used in the following examples
import chisel3._
import chisel3.util._
import chisel3.stage.ChiselStage
import chisel3.experimental.ChiselEnum
```

```

Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:16]: I had the import in separate ticks from the example. You removed the end ticks from "Importing the ChiselEnum module" but not the start ticks of the next block

Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:16]: yeah but i think that is an artifact of github's suggest changes API... anyway, consider this just as a suggestion

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:13]:
```suggestion
import chisel3.experimental.ChiselEnum
```

```

Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:13]: sigh. i swear i am suggesting the additional backticks!

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md]: something weird here in the formatting. You can actually make the `mdoc` just display the thrown error by doing the bad thing in the code

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:13]: I pushed a new commit with backticks

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md]: I added space around the block quote for now. I'm worried that the ci will start going crazy if I intentionally throw errors in the mdoc

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [inline docs/src/explanations/chisel-enum.md:28]: To narrowly document `ChiselEnum`, it's sufficient to ignore suggestions of how a user should organize/package their code.

That said, if you want to show this as living under `CPUTypes`, I think that's reasonable. Because Scala seemingly treats objects and packages as basically the same, you can show what you want with:
```scala
object CPUTypes {
  object AluMux1Sel extends Chisel Enum { /** */ }
}
```

Comment (seldridge) [reply docs/src/explanations/chisel-enum.md:28]: Additionally, it would be good to make clear what the `"selectRS1" -> 0.U` portion is supposed to be doing and what that does. It would be fine to show this with two different variants of `AluMux1Sel`, e.g., `AluMux1Sel` and `AluMex1SelExplicitValues` or something.

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [inline docs/src/explanations/chisel-enum.md:80]: Any code block should use the `scala mdoc` tag. This this to get type checked by the compiler and ensure that our documentation isn't heinously busted. 😄

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [inline docs/src/explanations/chisel-enum.md:84]:
```suggestion
```scala mdoc
```

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [inline docs/src/explanations/chisel-enum.md:104]:
```suggestion
```scala mdoc
```

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [inline docs/src/explanations/chisel-enum.md:118]:
```suggestion
```scala mdoc
```

**Review by seldridge — COMMENTED** [review]
Thanks for your first PR! 👏 I really appreciate you taking the time to write this up.

This is looking good and I'll leave it to @mwachs5 to continue handling the actual content review.

My only real concern is that everything needs to run through `mdoc` so that it's type checked. Use the `scala mdoc` tag instead of a bare `scala` tag. 

Two FYI comments:
- You can build this locally to prevent having to wait on CI with `sbt docs/mdoc` (or keep an `sbt` shell open and just run `docs/mdoc` in it). The output, post-`mdoc` processing shows up in `docs/generated`.
- There are a lot of useful `mdoc` modifiers documented [here](https://scalameta.org/mdoc/docs/modifiers.html). This includes things for code which is expected to error.

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [reply docs/src/explanations/chisel-enum.md]: For an expected exception, I think `scala mdoc:crash` is what you want. There's also `scala mdoc:fail`. Lots of documentation on this [here](https://scalameta.org/mdoc/docs/modifiers.html).

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:68]:
```suggestion
```scala mdoc
```

File: docs/src/explanations/chisel-enum.md
Comment (chrisbaldwin2) [reply docs/src/explanations/chisel-enum.md:28]: Right so this the mapping comment comes from a misunderstanding I had with the value assignment/verification code in the example StrongEnum code. This is the actual code that I have in my file

```scala
object AluMux1Sel extends ChiselEnum {
        val rs1out, pcout = Value
        val correct_annotation_map = Map[String, UInt](elems =
            "rs1out" -> 0.U,
            "pcout"  -> 1.U
        )
    }
```

When I first saw this mapping function, I thought the mapping is what actually assigned the value BUT IT DOESN'T!!! For the documentation I removed the actual function because it is only for verification and not for assigning value.

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [inline docs/src/explanations/chisel-enum.md:134]:
```suggestion
The ChiselEnum type is much more powerful than stated above. It allows for Sequence, Vec, and Bundle assignments, as well as a `.next` operation to allow for stepping through sequential states and an `.isValid` for checking that a hardware value is a valid `Value`. The source code for the ChiselEnum can be found [here](https://github.com/chipsalliance/chisel3/blob/2a96767097264eade18ff26e1d8bce192383a190/core/src/main/scala/chisel3/StrongEnum.scala) in the class `EnumFactory`. Examples of the ChiselEnum operations can be found [here](https://github.com/chipsalliance/chisel3/blob/dd6871b8b3f2619178c2a333d9d6083805d99e16/src/test/scala/chiselTests/StrongEnum.scala).
```

File: docs/src/explanations/chisel-enum.md
Comment (mwachs5) [reply docs/src/explanations/chisel-enum.md:28]: i think the current form of the doc is helpful, I found it clear what it was saying

**Comment by mwachs5** [comment]
@seldridge, @jackkoenig general question about this... it seems that since `ChiselEnum` is in the `experimental` package, it doesn't show up on the API docs? What's the general policy on including things from the `experimental`package in the API docs / the website documentation? It seems strange that we would have a nice doc like this but not an API doc. If the policy is to not put experimental features into the API docs, maybe we should have a docs division of "experimental features" to make this clear

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [reply docs/src/explanations/chisel-enum.md:28]: Ah, I get it now. Maybe just extend the comment to say: "This is what the mapping will be" or something. Not blocking. I'm fine with it the way it is. :+1:

**Review by seldridge — APPROVED** [review]

**Comment by mwachs5** [comment]
@chrisbaldwin2 can you please fill out the contributor checklist at the top? this is looking good!

**Comment by seldridge** [comment]
@mwachs5 wrote:
> general question about this... it seems that since ChiselEnum is in the experimental package, it doesn't show up on the API docs (https://www.chisel-lang.org/api/3.4.2/?search=ChiselEnum)? 

I'm not sure what's going on here, but it's not omitted intentionally. FWIW, it _does_ exist, but it appears that Scaladoc search doesn't work for type definitions? I.e., it exists here: https://www.chisel-lang.org/api/latest/chisel3/experimental/index.html#ChiselEnum=chisel3.experimental.EnumFactory, but I can't search for that.

The organization of the `ChiselEnum` stuff always seemed weird to me and worth a refactor... Why is it `type ChiselEnum = EnumFactory`? 

@mwachs5 wrote:
> What's the general policy on including things from the experimental package in the API docs / the website documentation? It seems strange that we would have a nice doc like this but not an API doc. If the policy is to not put experimental features into the API docs, maybe we should have a docs division of "experimental features" to make this clear

I think everything should be documented unless it's a non-public API that we're trying to hide. (The only thing in that category is anything `private [chisel3]` which should already be stripped and the public `do_foo` methods which use some magic to put them in a low priority group.) Experimental stuff does appear to be documented, but it's `ChiselEnum` that search doesn't like.

Adding API docs sounds like a good feature request. :+1:

File: docs/src/explanations/chisel-enum.md
Comment (seldridge) [inline docs/src/explanations/chisel-enum.md:22]: :100:

**Comment by chrisbaldwin2** [comment]
I updated the PR checklist information. I believe that this is ready to be merged! Thanks for your time and help! I hope that this documentation can lead to more people using this great feature of Chisel!!!

**Comment by mwachs5** [comment]
I don't have write access, but I approve. Thanks for doing this!

**Comment by seldridge** [comment]
Thanks again for the contribution @chrisbaldwin2. This will show up on the website after the Chisel 3.4.3 release goes out (not sure on ETA for that).

**Pull request merged**
